cli/azd/pkg/cache/file_cache.go

```diff
@@ -0,0 +1,98 @@
+package cache
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/azure/azure-dev/cli/azd/pkg/osutil"
+)
+
+// CacheResolver is a function that resolves the cache value.
+type CacheResolver[T any] func(ctx context.Context) (*T, error)
+
+// FileCache is a cache that stores the value in a file otherwise resolves it.
```

Comment:
Nice abstraction - I'll try to move the stuff that caches the latest version for our update check over to it if possible.

You might note that it's the JSON encoding of `*T` that is stored in the cache here, since that matters a bit for consumers.